
## R script to generate list of association matrices by class from raw data

## generates tables based on presence absence and then weighted

library(phyloseq)
library(ape)
library(bipartite )
library(bipartiteD3)
library(reshape2)
library(expss)
library(ggsci)
library(tidyverse)
library(metagMisc)

### IMPORT DATA

plasmid_data_ps<-readRDS("DATA/plasmid_50pc_97pc_unmerged.RDS") #this is a phyloseq object with three levels of data

plasmid_data_ps # This is in phyloseq object format, three levels of data

#2770 plasmids
#191 bacteria

########## import taxonomy and phylogeny of the 191 taxa
taxonomy <- read.csv("DATA/taxonomy_phylophlan.csv", sep=",") #taxonomy for genome clusters generated by PhyloPhaln
tr<-read.tree("DATA/phylophlan2.tre.treefile") #phylo tree for genome clusters generated by PhyloPhlan

###########

#### add taxonomy only metadata
#### add taxonomy only metadata
#### add taxonomy only metadata
#### add taxonomy only metadata


sample_data(plasmid_data_ps)$Phylum<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Phylum")
sample_data(plasmid_data_ps)$Class<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Class")
sample_data(plasmid_data_ps)$Order<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Order")
sample_data(plasmid_data_ps)$Family<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Family")
sample_data(plasmid_data_ps)$Genus<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Genus")
sample_data(plasmid_data_ps)$Species<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Species")


################# filter out rare plasmid contigs ###############
################# filter out rare plasmid contigs ###############

phylo_filtered<-prune_taxa(taxa_sums(plasmid_data_ps) > 20, plasmid_data_ps) 



### look for the most common classes in the data

Class_freq<-data.frame(table(sample_data(phylo_filtered)$Class))
head(Class_freq)
Class_freq<-Class_freq[order(-Class_freq$Freq),]

head(Class_freq, 11)

## keep only classes with 8 or more taxa (can edit)

classes_to_keep<-subset(Class_freq, Freq > 4)
classes_to_keep<-as.character(classes_to_keep$Var1)


sample_data(phylo_filtered)$MajorClass<-sample_data(phylo_filtered)$Class %in% classes_to_keep
phylo_classes<-subset_samples(phylo_filtered, MajorClass == TRUE)

phylo_classes

## remove taxa that no longer occur

phylo_classes<-prune_taxa(taxa_sums(phylo_classes) > 0, phylo_classes) 

### now split phyloseq object into list of x different objects, by class

phylo_by_class<-metagMisc::phyloseq_sep_variable(phylo_classes, "Class", drop_zeroes = T)

#list of new object seperated by class
phylo_by_class

############################################## order #######
############################################## order #######
############################################## order #######
############################################## order #######

phylo_by_class<-phylo_by_class[c("c__Betaproteobacteria",  "c__Gammaproteobacteria", 
  "c__Bacteroidia",
  "c__Clostridia",
  "c__Actinobacteria")]


phylo_by_class$All_classes<-phylo_filtered

phylo_by_class<-phylo_by_class[c("All_classes", "c__Betaproteobacteria",  "c__Gammaproteobacteria", 
  "c__Bacteroidia",
  "c__Clostridia",
  "c__Actinobacteria")]

####################################################### make list of presence/absence association tables by class ####
####################################################### make list of presence/absence association tables by class ####
####################################################### make list of presence/absence association tables by class ####
####################################################### make list of presence/absence association tables by class ####
####################################################### make list of presence/absence association tables by class ####

uniq<-names(phylo_by_class)
assmat_by_class_pa<-list()


for (i in 1:length(uniq)){
  
  data<-phylo_by_class[[i]]
  data_pa <-phyloseq_standardize_otu_abundance(data, method = "pa")
  otutable<-data.frame(otu_table(data_pa))
  assmat_by_class_pa[[i]]<-otutable
}

names(assmat_by_class_pa)<-uniq
assmat_by_class_pa


############################ make list of weighted association tables by class ###
############################ make list of weighted association tables by class ###
############################ make list of weighted association tables by class ###
############################ make list of weighted association tables by class ###
############################ make list of weighted association tables by class ###

uniq<-names(phylo_by_class)
assmat_by_class_weighted<-list()


for (i in 1:length(uniq)){
  
  data<-phylo_by_class[[i]]
  otutable<-data.frame(otu_table(data))
  assmat_by_class_weighted[[i]]<-otutable
}

names(assmat_by_class_weighted)<-uniq
assmat_by_class_weighted

