
### script to make Figure 1a and b
### bipartite network plots

library(phyloseq)
library(ape)
library(bipartite )
library(bipartiteD3)
library(reshape2)
library(expss)
library(ggsci)
library(tidyverse)
library(metagMisc)

### IMPORT DATA

plasmid_data_ps<-readRDS("plasmid_50pc_97pc_unmerged.RDS") #this is a phyloseq object with three levels of data

plasmid_data_ps # This is in phyloseq object format, three levels of data

#2770 plasmids
#191 bacteria

########## import taxonomy and phylogeny of the 191 taxa
taxonomy <- read.csv("taxonomy_phylophlan.csv", sep=",") #taxonomy for genome clusters generated by PhyloPhaln
tr<-read.tree("phylophlan2.tre.treefile") #phylo tree for genome clusters generated by PhyloPhlan

###########

#### add taxonomy only metadata
#### add taxonomy only metadata
#### add taxonomy only metadata
#### add taxonomy only metadata


sample_data(plasmid_data_ps)$Phylum<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Phylum")
sample_data(plasmid_data_ps)$Class<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Class")
sample_data(plasmid_data_ps)$Order<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Order")
sample_data(plasmid_data_ps)$Family<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Family")
sample_data(plasmid_data_ps)$Genus<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Genus")
sample_data(plasmid_data_ps)$Species<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Species")


################# filter out rare plasmid contigs ###############
################# filter out rare plasmid contigs ###############

phylo_filtered<-prune_taxa(taxa_sums(plasmid_data_ps) > 20, plasmid_data_ps) 


############### FIGURE 1A ###########################
############### FIGURE 1A ###########################
############### FIGURE 1A ###########################
############### FIGURE 1A ###########################

### generate long data table of interactions

otu_table<-data.frame(phylo_filtered@otu_table@.Data)

### long format 

otu_matrix<-as.matrix(otu_table)
row.names(otu_matrix)<-row.names(otu_table)
otu_long<-melt(otu_matrix, na.rm = T)
otu_long<-subset(otu_long, value !=0) # remove zeros
names(otu_long)<-c("lower", "higher", "freq") #rename cols
head(otu_long)

### get taxonomy

taxtable<-data.frame(sample_data(phylo_filtered))[,c(1, 20)]
head(taxtable)

####

## 
otu_long$webID<-vlookup(otu_long$lower, taxtable, lookup_column = "feature.id", result_column = "Class")
head(otu_long)

#### only keep main classes which have over 15 associations
Class_freq<-data.frame(table(otu_long$webID))
Class_freq<-Class_freq[order(-Class_freq$Freq),]
head(Class_freq, 15)

## keep only classes with 8 or more taxa (can edit)
classes_to_keep<-subset(Class_freq, Freq > 5)
classes_to_keep<-as.character(classes_to_keep$Var1)
otu_long$MajorClass<-otu_long$webID %in% classes_to_keep
otu_long_reduced<-subset(otu_long, MajorClass == TRUE)
head(otu_long_reduced)
otu_long_reduced<-otu_long_reduced[,-5]

unique(otu_long_reduced$webID)

otu_long_reduced$webID<-factor(otu_long_reduced$webID)

head(otu_long_reduced)
names(otu_long_reduced)[4]<-"Class"
otu_long_reduced$webID<-"all"

############################### plotting 
############################### plotting 
############################### plotting 
############################### plotting 

## get in right format for plotting
bipartite::frame2webs(otu_long_reduced)-> plasmid_network_all

## need genrate vector of plasmid and bacteria order for plotting
# bacteria should be ordered by taxonomy and then number of associations
#plasmids should be ordered by the class to which they are majorly associated with, 
#and then by numebr of associations

df<-bipartiteD3::List2DF(plasmid_network_all)
head(df)
#Primary = bacteria
#Secondary = Plasmids

df$Class<-vlookup(df$Primary, taxtable, lookup_column = "feature.id", result_column = "Class")
df$Class<-factor(df$Class)
unique(df$Class)
head(df)

# To sort secondary/plasmids by  bacteria class they are mostly associatied with and total size:
df %>%
  group_by(Secondary, Class) %>%
  summarise(Total=sum(all))-> SortDf_s

SortDf_s2<-SortDf_s %>% group_by(Secondary) %>% top_n(1, Total) 
SortDf_s2$Class<-factor(SortDf_s2$Class)
unique(SortDf_s2$Class)

SortDf_s2$Class<-factor(SortDf_s2$Class, levels = c("c__Betaproteobacteria",  "c__Gammaproteobacteria", "c__Epsilonproteobacteria", "c__Alphaproteobacteria", 
  "c__Bacteroidia",  "c__Actinobacteria", "c__Clostridia", "c__Bacilli",  "c__Fusobacteriia", "c__Caldilineae", "c__Nitrospira"))

SortDf_s2 %>% arrange(Class,desc(Total))-> SortDf_s2

### sort bacteria/primary by class and number of associations

df %>%
  group_by(Primary) %>%
  summarise(Total=sum(all))-> SortDf_p

SortDf_p$Class<-vlookup(SortDf_p$Primary, taxtable, lookup_column = "feature.id", result_column = "Class")
SortDf_p$Class<-factor(SortDf_p$Class)
unique(SortDf_p$Class)


SortDf_p$Class<-factor(SortDf_p$Class, levels = c("c__Betaproteobacteria",  "c__Gammaproteobacteria", "c__Epsilonproteobacteria", "c__Alphaproteobacteria", 
  "c__Bacteroidia", "c__Flavobacteriia", "c__Actinobacteria", "c__Clostridia", "c__Bacilli","c__Negativicutes",  "c__Fusobacteriia", "c__Caldilineae", "c__Nitrospira"))

SortDf_p  %>% arrange(Class,desc(Total))-> SortDf_p


### colour by class

head(otu_long_reduced)
unique(otu_long_reduced$Class)

mypal =  	pal_jco("default", alpha = 1)(8)
mypal1 =  	pal_locuszoom("default", alpha = 1)(8)
mypal2 =  	pal_npg("nrc", alpha = 1)(8)
mypal3 =  	pal_uchicago("dark", alpha = 1)(8)

### set colours 
#proteobacteria
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Betaproteobacteria", mypal[1], NA)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Gammaproteobacteria", mypal[6], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Epsilonproteobacteria",  mypal1[4], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Alphaproteobacteria",  "gray", otu_long_reduced$colour)

#bacteroidetes
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Bacteroidia",  mypal2[3], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Flavobacteriia",  "forestgreen", otu_long_reduced$colour)

#firmicutes
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Clostridia",  mypal2[8], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Bacilli",  mypal2[5], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Negativicutes",  mypal3[1], otu_long_reduced$colour)

#actinobacteria and fusobacteria
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Actinobacteria",  mypal[2], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Fusobacteriia",  mypal1[6], otu_long_reduced$colour)

otu_long_reduced$colour<-ifelse(is.na(otu_long_reduced$colour),  "gray", otu_long_reduced$colour)

#### generate vector of colours

ManualColours1<-otu_long_reduced[,c(1,6)]
ManualColours1<-distinct(ManualColours1)

row.names(ManualColours1)<-ManualColours1$lower
head(ManualColours1)

ManualColours2<-ManualColours1[,2]
names(ManualColours2)<-ManualColours1$lower

######################plot 
######################plot 
######################plot 
######################plot 

##interactive

bipartite_D3(plasmid_network_all, 
  SortSecondary = SortDf_s2$Secondary, 
  SortPrimary = SortDf_p$Primary,
  colouroption = 'manual',
  NamedColourVector = ManualColours2, 
  ColourBy = 1,
  MainFigSize = c(3000, 1200), 
  IndivFigSize = c(600, 1700),
  #BoxLabPos = c(20, 20),
  #PercPos = c(200,200),
  Pad = 0.7,
  BarSize = 120,
  MinWidth = 0.5,
  PercentageDecimals = 1,
  Orientation = 'horizontal',
  filename = 'all_classes')

## save by downloading pdf/png from plot screen since its interactive


######################## FIGURE 1B PRESENCE/ABSENCE ############


phylo_filtered<-prune_taxa(taxa_sums(plasmid_data_ps) > 20, plasmid_data_ps) 
phylo_transformed <-metagMisc::phyloseq_standardize_otu_abundance(phylo_filtered, method = "pa")

### generate long data table of interactions

otu_table<-data.frame(phylo_transformed@otu_table@.Data)

### long format 

otu_matrix<-as.matrix(otu_table)
row.names(otu_matrix)<-row.names(otu_table)
otu_long<-melt(otu_matrix, na.rm = T)
otu_long<-subset(otu_long, value !=0) # remove zeros
names(otu_long)<-c("lower", "higher", "freq") #rename cols
head(otu_long)

### get taxonomy

taxtable<-data.frame(sample_data(phylo_transformed))[,c(1, 20)]
head(taxtable)

####

## 
otu_long$webID<-vlookup(otu_long$lower, taxtable, lookup_column = "feature.id", result_column = "Class")
head(otu_long)

#### only keep main classes which have over 15 associations
Class_freq<-data.frame(table(otu_long$webID))
Class_freq<-Class_freq[order(-Class_freq$Freq),]
head(Class_freq, 15)

## keep only classes with 8 or more taxa (can edit)
classes_to_keep<-subset(Class_freq, Freq > 5)
classes_to_keep<-as.character(classes_to_keep$Var1)
otu_long$MajorClass<-otu_long$webID %in% classes_to_keep
otu_long_reduced<-subset(otu_long, MajorClass == TRUE)
head(otu_long_reduced)
otu_long_reduced<-otu_long_reduced[,-5]

unique(otu_long_reduced$webID)

otu_long_reduced$webID<-factor(otu_long_reduced$webID)

head(otu_long_reduced)
names(otu_long_reduced)[4]<-"Class"
otu_long_reduced$webID<-"all"

############################### plotting 
############################### plotting 
############################### plotting 
############################### plotting 

## get in right format for plotting
bipartite::frame2webs(otu_long_reduced)-> plasmid_network_all

## need genrate vector of plasmid and bacteria order for plotting
# bacteria should be ordered by taxonomy and then number of associations
#plasmids should be ordered by the class to which they are majorly associated with, 
#and then by numebr of associations

df<-bipartiteD3::List2DF(plasmid_network_all)
head(df)
#Primary = bacteria
#Secondary = Plasmids

df$Class<-vlookup(df$Primary, taxtable, lookup_column = "feature.id", result_column = "Class")
df$Class<-factor(df$Class)
unique(df$Class)
head(df)

# To sort secondary/plasmids by  bacteria class they are mostly associatied with and total size:
df %>%
  group_by(Secondary, Class) %>%
  summarise(Total=sum(all))-> SortDf_s

SortDf_s2<-SortDf_s %>% group_by(Secondary) %>% top_n(1, Total) 
SortDf_s2$Class<-factor(SortDf_s2$Class)
unique(SortDf_s2$Class)

SortDf_s2$Class<-factor(SortDf_s2$Class, levels = c("c__Betaproteobacteria",  "c__Gammaproteobacteria", "c__Epsilonproteobacteria", "c__Alphaproteobacteria", 
  "c__Bacteroidia",  "c__Actinobacteria", "c__Clostridia", "c__Bacilli",  "c__Fusobacteriia", "c__Caldilineae", "c__Nitrospira"))

SortDf_s2 %>% arrange(Class,desc(Total))-> SortDf_s2

### sort bacteria/primary by class and number of associations

df %>%
  group_by(Primary) %>%
  summarise(Total=sum(all))-> SortDf_p

SortDf_p$Class<-vlookup(SortDf_p$Primary, taxtable, lookup_column = "feature.id", result_column = "Class")
SortDf_p$Class<-factor(SortDf_p$Class)
unique(SortDf_p$Class)


SortDf_p$Class<-factor(SortDf_p$Class, levels = c("c__Betaproteobacteria",  "c__Gammaproteobacteria", "c__Epsilonproteobacteria", "c__Alphaproteobacteria", 
  "c__Bacteroidia", "c__Flavobacteriia", "c__Actinobacteria", "c__Clostridia", "c__Bacilli","c__Negativicutes",  "c__Fusobacteriia", "c__Caldilineae", "c__Nitrospira"))

SortDf_p  %>% arrange(Class,desc(Total))-> SortDf_p


### colour by class

head(otu_long_reduced)
unique(otu_long_reduced$Class)

mypal =  	pal_jco("default", alpha = 1)(8)
mypal1 =  	pal_locuszoom("default", alpha = 1)(8)
mypal2 =  	pal_npg("nrc", alpha = 1)(8)
mypal3 =  	pal_uchicago("dark", alpha = 1)(8)

### set colours 
#proteobacteria
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Betaproteobacteria", mypal[1], NA)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Gammaproteobacteria", mypal[6], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Epsilonproteobacteria",  mypal1[4], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Alphaproteobacteria",  "gray", otu_long_reduced$colour)

#bacteroidetes
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Bacteroidia",  mypal2[3], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Flavobacteriia",  "forestgreen", otu_long_reduced$colour)

#firmicutes
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Clostridia",  mypal2[8], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Bacilli",  mypal2[5], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Negativicutes",  mypal3[1], otu_long_reduced$colour)

#actinobacteria and fusobacteria
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Actinobacteria",  mypal[2], otu_long_reduced$colour)
otu_long_reduced$colour<-ifelse(otu_long_reduced$Class == "c__Fusobacteriia",  mypal1[6], otu_long_reduced$colour)

otu_long_reduced$colour<-ifelse(is.na(otu_long_reduced$colour),  "gray", otu_long_reduced$colour)

#### generate vector of colours

ManualColours1<-otu_long_reduced[,c(1,6)]
ManualColours1<-distinct(ManualColours1)

row.names(ManualColours1)<-ManualColours1$lower
head(ManualColours1)

ManualColours2<-ManualColours1[,2]
names(ManualColours2)<-ManualColours1$lower

######################plot 
######################plot 
######################plot 
######################plot 

##interactive

bipartite_D3(plasmid_network_all, 
  SortSecondary = SortDf_s2$Secondary, 
  SortPrimary = SortDf_p$Primary,
  colouroption = 'manual',
  NamedColourVector = ManualColours2, 
  ColourBy = 1,
  MainFigSize = c(3000, 1200), 
  IndivFigSize = c(600, 1700),
  #BoxLabPos = c(20, 20),
  #PercPos = c(200,200),
  Pad = 0.7,
  BarSize = 120,
  MinWidth = 0.5,
  PercentageDecimals = 1,
  Orientation = 'horizontal',
  filename = 'all_classes')

## save by downloading pdf/png from plot screen since its interactive


