

# This code generates a dataframe of overall network stats per class (connectance, links per species, compartment diversity etc), 
#and generates and plots each network

library(phyloseq)
library(ape)
library(bipartite )
library(bipartiteD3)
library(reshape2)
library(expss)
library(ggsci)
library(tidyverse)
library(metagMisc)
library(igraph)
library(network)
library(intergraph)
library(scales)
library(qgraph)

### IMPORT DATA

plasmid_data_ps<-readRDS("DATA/plasmid_50pc_97pc_unmerged.RDS") #this is a phyloseq object with three levels of data

plasmid_data_ps # This is in phyloseq object format, three levels of data

#2770 plasmids
#191 bacteria

########## import taxonomy and phylogeny of the 191 taxa
taxonomy <- read.csv("DATA/taxonomy_phylophlan.csv", sep=",") #taxonomy for genome clusters generated by PhyloPhaln
tr<-read.tree("DATA/phylophlan2.tre.treefile") #phylo tree for genome clusters generated by PhyloPhlan

###########

#### add taxonomy only metadata
#### add taxonomy only metadata
#### add taxonomy only metadata
#### add taxonomy only metadata


sample_data(plasmid_data_ps)$Phylum<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Phylum")
sample_data(plasmid_data_ps)$Class<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Class")
sample_data(plasmid_data_ps)$Order<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Order")
sample_data(plasmid_data_ps)$Family<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Family")
sample_data(plasmid_data_ps)$Genus<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Genus")
sample_data(plasmid_data_ps)$Species<-vlookup(sample_data(plasmid_data_ps)$feature.id, taxonomy, lookup_column = "cluster_id", result_column = "Species")


################# filter out rare plasmid contigs ###############
################# filter out rare plasmid contigs ###############

phylo_filtered<-prune_taxa(taxa_sums(plasmid_data_ps) > 20, plasmid_data_ps) 



### look for the most common classes in the data

Class_freq<-data.frame(table(sample_data(phylo_filtered)$Class))
head(Class_freq)
Class_freq<-Class_freq[order(-Class_freq$Freq),]

head(Class_freq, 11)

## keep only classes with 8 or more taxa (can edit)

classes_to_keep<-subset(Class_freq, Freq > 4)
classes_to_keep<-as.character(classes_to_keep$Var1)


sample_data(phylo_filtered)$MajorClass<-sample_data(phylo_filtered)$Class %in% classes_to_keep
phylo_classes<-subset_samples(phylo_filtered, MajorClass == TRUE)

phylo_classes

## remove taxa that no longer occur

phylo_classes<-prune_taxa(taxa_sums(phylo_classes) > 0, phylo_classes) 

### now split phyloseq object into list of x different objects, by class

phylo_by_class<-metagMisc::phyloseq_sep_variable(phylo_classes, "Class", drop_zeroes = T)

#list of new object seperated by class
phylo_by_class


########### networks and network stats by class #####
########### networks and network stats by class #####
########### networks and network stats by class #####
########### networks and network stats by class #####



network_stats_list<-list()

uniq<-names(phylo_by_class)

#par(mfrow=c(2,2))

for (i in 1:length(uniq)){
  
  phylo<- phylo_by_class[[i]]
  #phylo<- phylo_by_class$c__Actinobacteria
  
  net.v<-data.frame(phylo@otu_table@.Data)
  
  net.v[1:5, 1:5]
  
  # order network
  net.v2<-sortweb(net.v, sort.order="dec", sequence=NULL)
  
  # PDI index 
  spec.gen<-PDI(net.v2, normalise=TRUE, log=TRUE)#Returns a vector with PDI values between 0 (perfect generalist) and 1 (perfect specialist).
  # histogram(spec.gen)
  
  # this will estimate all common network metrics but may take a while
  network_metrics<-data.frame(networklevel(net.v))
  network_metrics$Class<- sample_data(phylo)$Class[1]
  network_metrics$Phylum<- sample_data(phylo)$Phylum[1]
  network_metrics$Metric<-row.names(network_metrics)
  network_metrics<-network_metrics[,c(4,1,2,3)]
  names(network_metrics)[2]<-"Stat"
  
  head(network_metrics)
  network_stats_list[[i]]<-network_metrics
}



network_stats_df<-do.call(rbind, network_stats_list)

head(network_stats_df)
#write.csv(network_stats_df, "network_stats_per_class.csv")


############################ plot networks
############################ plot networks
############################ plot networks
############################ plot networks
############################ plot networks

uniq<-names(phylo_by_class)

net_list<-list()

par(mfrow=c(2,3))

for (i in 1:length(uniq)){
  
  phylo<- phylo_by_class[[i]]
  #phylo<- phylo_by_class$c__Actinobacteria
  
  otu_table<-data.frame(phylo@otu_table@.Data)
  otu_matrix<-as.matrix(otu_table)
  row.names(otu_matrix)<-row.names(otu_table)
  otu_long<-melt(otu_matrix, na.rm = T)
  otu_long<-subset(otu_long, value !=0)
  head(otu_long)
  names(otu_long)<-c("host", "plasmid", "count")
  
  
  ### nodes
  
  sources <- otu_long %>%
    distinct(host) %>%
    rename(label = host)
  
  destinations <- otu_long%>%
    distinct(plasmid) %>%
    rename(label = plasmid)
  
  nodes <- full_join(sources, destinations, by = "label")
  head(nodes)
  
  nodes <- nodes %>% rowid_to_column("id")
  nodes$label<-as.character(nodes$label)
  
  
  ##edges
  
  per_route <- otu_long[,c(1:3)]
  head(per_route)
  names(per_route)[3]<-"weight" 
  
  edges <- per_route %>% 
    left_join(nodes, by = c("host" = "label")) %>% 
    rename(from = id)
  
  edges <- edges %>% 
    left_join(nodes, by = c("plasmid" = "label")) %>% 
    rename(to = id)
  head(edges)
  
  edges <- select(edges, from, to, weight)
  head(edges)
  
  #################### make into network object
  
  
  routes_network <- network(edges, vertex.attr = nodes, matrix.type = "edgelist", ignore.eval = FALSE)
  igraph_net<-intergraph::asIgraph(routes_network)
  igraph_net<-as.undirected(igraph_net)
  
  ### add metadata
  
  V(igraph_net)$contig_id<-vlookup(V(igraph_net)$id, nodes, lookup_column = "id", result_column = "label")
  V(igraph_net)$Type<-ifelse(V(igraph_net)$contig_id %in% sources$label, "Host", "Plasmid")
  table(V(igraph_net)$Type)
  
  ##### add network stats per node
  
  net.cn <- igraph::closeness(igraph_net) #closeness centrality
  net.bn <- igraph::betweenness(igraph_net) #betweenness centrality
  net.deg<-igraph::degree(igraph_net) #degree
  net.hs <- igraph::hub_score(igraph_net)$vector
  
  
  V(igraph_net)$closeness<-net.cn
  V(igraph_net)$betweenness<-net.bn
  V(igraph_net)$degree<-net.deg
  V(igraph_net)$hubbiness<-net.hs
  V(igraph_net)$contig_id_deg<-ifelse(V(igraph_net)$degree > 9, V(igraph_net)$contig_id, NA)
  
  ### add taxonomy 
  
  taxonomy<-data.frame(sample_data(phylo))
  V(igraph_net)$class<-as.character(vlookup(V(igraph_net)$contig_id, taxonomy, lookup_column = "feature.id", result_column = "Class"))
  
  ########## rescale weight
  
  E(igraph_net)$weight1<-log10(E(igraph_net)$weight)
  E(igraph_net)$weight2<-rescale(E(igraph_net)$weight1, to = c(0, 1))
  
  
  
  # Get clusters
  wt <- walktrap.community(igraph_net)
  cluster.membership<-membership(wt)
  nodes$ClusterMembership<-membership(wt)
  V(igraph_net)$ClusterMembership<-membership(wt)
  
  
  ############ set colour and shapes 
  ############ set colour and shapes 
  ############ set colour and shapes 
  ############ set colour and shapes 
  
  mypal =  	pal_jco("default", alpha = 1)(8)
  mypal1 =  	pal_locuszoom("default", alpha = 1)(8)
  mypal2 =  	pal_npg("nrc", alpha = 1)(8)
  mypal3 =  	pal_uchicago("dark", alpha = 1)(8)
  
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Betaproteobacteria", mypal[1], NA)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Gammaproteobacteria", mypal[6], V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Epsilonproteobacteria",  mypal1[4], V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Alphaproteobacteria",  "gray", V(igraph_net)$colour)
  
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Bacteroidia",  mypal2[3], V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Flavobacteriia",  "forestgreen", V(igraph_net)$colour)
  
  
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Clostridia",  mypal2[8], V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Bacilli",  mypal2[5], V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Negativicutes",  mypal3[1], V(igraph_net)$colour)
  
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Actinobacteria",  mypal[2], V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$class == "c__Fusobacteriia",  mypal1[6], V(igraph_net)$colour)
  
  V(igraph_net)$colour<-ifelse(is.na(V(igraph_net)$colour),  "gray", V(igraph_net)$colour)
  V(igraph_net)$colour<-ifelse(V(igraph_net)$Type=="Plasmid",  "white", V(igraph_net)$colour)
  
  ##################################
  ##################################
  
  par(mar=c(0,0,2,0)+0.1)
  
  plot(igraph_net,
    layout=layout_with_dh,
    vertex.size = (V(igraph_net)$degree/3) +3, 
    vertex.label = NA, 
    vertex.color = V(igraph_net)$colour,
    edge.width=(E(igraph_net)$weight/100)+1,
    edge.color = "gray")
  title(paste(uniq[i]),cex.main=2)
  
  ########################### clusetrs ##########
  ########################### clusetrs ##########
  ########################### clusetrs ##########
  ########################### clusetrs ##########
  ########################### clusetrs ##########
  ########################### clusetrs ##########
  #########
  
  
  ##### scale colour of edges to strength
  
  c_scale <- colorRamp(c('white','skyblue', 'darkblue')) # COLOUR SCALE
  E(igraph_net)$color <- apply(c_scale(E(igraph_net)$weight2), 1, function(x) rgb(x[1]/255,x[2]/255,x[3]/255) ) # ADJUST COLOURS
  
  
  ########################################
  
  
  plot.net.cls <- function(net, scores, cls, outfile, title) {
    # Get size of clusters to find isolated nodes.
    cls_sizes <- sapply(groups(cls), length)
    # Randomly choosing node colors. Users can provide their own vector of colors.
    colors <-c("forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "lightskyblue", "cyan","skyblue4", "skyblue3","darkslategray4","blue",
      "forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "forestgreen", "darkorange","bisque","purple","turquoise", "grey", "deeppink", "plum1","skyblue","dodgerblue","navyblue", 
      "green","cyan","darkred","brown","darkolivegreen","red","white")
    # Nodes in clusters will be color coded. Isolated nodes will be white.
    V(net)$color <- sapply(membership(cls),
      function(x) {ifelse(cls_sizes[x]>1,
        colors[x], "white")})
    
    
    # Define output image file.
    outfile <- paste(outfile, "jpg", sep=".")
    # Image properties.
    jpeg(paste("PLOTS/",outfile, sep = ""), width = 5000, height = 5000, res = 300, quality = 100)
    par(oma = c(1, 1, 1, 1))
    
    # Customized layout to avoid nodes overlapping.
    e <- get.edgelist(net)
    class(e) <- "numeric"
    l <- qgraph.layout.fruchtermanreingold(e, vcount=vcount(net),
      area=8*(vcount(net)^2),
      repulse.rad=(vcount(net)^3))
    
    # Main plot function.
    plot(net, vertex.size = (scores/4)+4,
      vertex.label=NA,
      vertex.label.dist = 0.6,
      vertex.label.cex = 2,
      vertex.label.font = 2,
      vertex.color=V(net)$colour,
      # vertex.shape=my_shape,
      vertex.label.color = "black",
      vertex.frame.color = "black",
      # mark.border= my_color,
      mark.expand = 10,
      mark.shape = 1,
      layout=l,
      edge.width=E(net)$weight/300+2,
      edge.color="darkgrey")
    title(title, cex.main=4)
    
    
    dev.off()
  }
  
  # Execute this command after running Function 3
  plot.net.cls(igraph_net, net.deg, wt,
    outfile = paste("plot_", uniq[i]), title = paste(uniq[i]))
  
  
  
  ### add network to list 
  
  net_list[[i]]<-igraph_net
  
}

